### ISSUE TITLE
"Specified group generation id is not valid" while running long duration consumer

### ISSUE THREAD

**Issue by rachits93** [confluentinc/confluent-kafka-dotnet#1155]
Description
===========
Hi Everyone,
I have 4 consumers running in a consumer group. The requirement is for each consumer to be able to process each message for long durations without any problems.

As part of load tests, i am making each consumer sleep for 30 mins after it receives a message ( as that is the worst case scenario for us), and then committing the offset synchronously after the 30 mins interval.

While making the commit after the 30 mins interval - this exception was thrown - 

Confluent.Kafka.KafkaException
  HResult=0x80131500
  Message=Broker: Specified group generation id is not valid
  Source=Confluent.Kafka
  StackTrace:
   at Confluent.Kafka.Impl.SafeKafkaHandle.Commit(IEnumerable`1 offsets)
   at Confluent.Kafka.Consumer`2.Commit(ConsumeResult`2 result)

Below is the code for consumer config - 

            var conf = new Dictionary<string, string>
            {
                { "group.id", _kafkaConsumerGroupName },
                { "enable.auto.commit", false.ToString() },
                { "auto.commit.interval.ms", "1000" },
                {"enable.auto.offset.store", false.ToString() },
                { "statistics.interval.ms", "60000" },
                { "bootstrap.servers", _kafkaBootstrapServers },
                { "max.poll.interval.ms", "3600000" }
            };

            var consumerConfig = new ConsumerConfig(conf);

            return consumerConfig;

Consumer is created like like below -

            var consumerBuidler = new ConsumerBuilder<TKafkaKey, string>(config)
                                  .SetLogHandler((consumer, logMessage) => 
                                  {
                                      _logMessageAction?.Invoke($"KafkaLog: {JsonConvert.SerializeObject(logMessage)}, consumer: {JsonConvert.SerializeObject(consumer)}");
                                  })
                                  .SetValueDeserializer(Deserializers.Utf8)
                                  .SetErrorHandler((consumer, error) =>
                                  {
                                      _logMessageAction?.Invoke($"KafkaLog: KafkaError: {JsonConvert.SerializeObject(error)}, Consumer: {JsonConvert.SerializeObject(consumer)}");
                                  })
                                  .SetPartitionsAssignedHandler((consumer, partitions) =>
                                  {
                                      _logMessageAction?.Invoke($"KafkaLog: Assigned partitions: {JsonConvert.SerializeObject(partitions)}, Consumer: {JsonConvert.SerializeObject(consumer)}");
                                  })
                                  .SetPartitionsRevokedHandler((consumer, partitions) =>
                                  {
                                      _logMessageAction?.Invoke($"KafkaLog: Revoked partitions: {JsonConvert.SerializeObject(partitions)}, Consumer: {JsonConvert.SerializeObject(consumer)}");
                                  });

            //Depending on the type passed in for TKafkaKey, create a consumer with the appropriate key deserializer
            if (typeof(TKafkaKey) == typeof(Null))
            {
                consumerBuidler = consumerBuidler.SetKeyDeserializer(Deserializers.Null as IDeserializer<TKafkaKey>);
            }
            else if (typeof(TKafkaKey) == typeof(string))
            {
                consumerBuidler = consumerBuidler.SetKeyDeserializer(Deserializers.Utf8 as IDeserializer<TKafkaKey>);
            }
            else
            {
                // changed from Exception to BindingException (that ActorAPI handles) with severity = FATAL to stop the Actor.
                throw new Exception("KafkaError: An invalid type was specified for the Kafka consumer Key. Only \"Null\" (Confluent.Kafka.Null) and \"String\" are allowed");
            }

            _kafkaConsumer = consumerBuidler.Build();

            //Subscribe
            _kafkaConsumer.Subscribe(_kafkaTopics);

After the initialization, below code is called for getting a message from the topic - 

            var result = _kafkaConsumer.Consume(TimeSpan.FromMilliseconds(_consumeTimeoutMS));

            if (result != null && result.Message != null)
            {
                _logMessageAction?.Invoke($"KafkaLog: ConsumeResult:{JsonConvert.SerializeObject(result)}");
            }

            return result;

The caller wait for 30 mins. Then commits the message like below - 

            _kafkaConsumer.Commit(result);

After committing the message, the calling app will fetch the message again and wait for 30 mins and then commit. This cycle is supposed to continue indefinetely. However, exception ( as mentioned at the top) was thrown when message was sent to be committed after the 30 mins interval. 

The code is working fine with message interval of 5 mins (instead of 30 mins). 

The max.poll.interval.ms has been set with a value of 1 hour. So i don't get why the duration is a problem. All other config values , apart from those mentioned in the code snippet, are at default.

I am using Confluent.Kafka Version 1.3.0

**Comment by mhowlett** [comment]
what was the issue? (i was just trying to figure this out as you closed it)

**Comment by rachits93** [comment]
I only recently saw the follow up question. 

> what was the issue? (i was just trying to figure this out as you closed it)

I don't remember exactly what had been to done to fix it, because it has been months now. However, i think it might have something to do with session.timeout.ms value. When i created this post, i was using its default value (which i believe is 10 seconds), however i increased it later to 5 mins. This may have fixed the issue for me. I also adjusted the heartbeat interval timeout value (since its supposed to be 1/3 rd or less than 1/3rd of the session timeout).  Rest of the consumer config values are as in the original post.

Hope this helps.
